File: java/templatemethod/CVReportGeneration.java
package templatemethod;

import java.io.File;

public abstract class CVReportGeneration {

    public GeneratedReport generateCVReport(String cvFilePath) {
        var file = readFile(cvFilePath);
        var extractedData = extractData(file);
        var analyzedData = analyzeData(extractedData);
        return generateReportOf(analyzedData);
    }

    protected abstract Data extractData(File file);

    private File readFile(String filePath) {
        System.out.println("reading file from: " + filePath);
        return null;
    }

    private AnalyzedData analyzeData(Data data) {
        System.out.println("analyzing data...");
        return null;
    }

    private GeneratedReport generateReportOf(AnalyzedData analyzedData) {
        System.out.println("generating report...");
        return new GeneratedReport(true);
    }
}

File: C#/TemplateMethod/CvReportGeneration.cs
using System.Net;
using dotnet.TemplateMethod.Models;

namespace dotnet.TemplateMethod;

public abstract class CvReportGeneration
{
    public GeneratedReport GenerateReport(Uri cvPath)
    {
        var fileBuffer = ReadFile(cvPath);
        var data = ExtractData(cvPath);
        var analyzedData = AnalyzeData(data);
        return GenerateReport(analyzedData);
    }
    protected abstract Data ExtractData(Uri cvPath);
    
    private byte[] ReadFile(Uri cvPath)
    {
        Console.WriteLine("Reading file...");
        
        return Array.Empty<byte>();
    }
    
    private AnalyzedData AnalyzeData(Data data)
    {
        Console.WriteLine("Analyzing data...");
        
        return new AnalyzedData();
    }
    
    private GeneratedReport GenerateReport(AnalyzedData analyzedData)
    {
        Console.WriteLine("Generating report...");
        
        return new GeneratedReport(true);
    }
}